server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java

```diff
@@ -716,7 +749,7 @@ public Index(StreamInput in) throws IOException {
             int size = in.readVInt();
             for (int i = 0; i < size; i++) {
                 File file = new File(in);
-                fileDetails.put(file.name, file);
+                fileDetails.addFileDetails(file.name, file);
```

Comment:
Think it'd be better to move this to a constructor `RecoveryFilesDetails(StreamInput in)`, and similarly implement `RecoveryFileDetails#writeTo` instead of implementing the serialization here.